zeppelin-interpreter/src/test/java/org/apache/zeppelin/interpreter/ZeppCtxtVariableTest.java

```diff
@@ -0,0 +1,184 @@
+    String result = interpreter.interpolate("The value of {{PI}} is {PI} now", resourcePool);
+    assertTrue("Substitution in the middle", "The value of {PI} is 3.1415 now".equals(result));
+  }
+
+  @Test
+  public void substitutionAtTheEnds() {
+    String result = interpreter.interpolate("{{PI}} is now {PI}", resourcePool);
+    assertTrue("Substitution at the ends", "{PI} is now 3.1415".equals(result));
+  }
+
+  @Test
+  public void multiLineSubstitution() {
+    String result = interpreter.interpolate("{{PI}}\n{PI}\n{{PI}}\n{PI}", resourcePool);
+    assertTrue("Substitution at the ends", "{PI}\n3.1415\n{PI}\n3.1415".equals(result));
+  }
+
+  @Test
+  public void noUndefinedVariableError() {
+    String result = interpreter.interpolate("This {pi} will pass silently", resourcePool);
+    assertTrue("No partial substitution", "This {pi} will pass silently".equals(result));
+  }
+
+  @Test
+  public void noPartialSubstitution() {
+    String result = interpreter.interpolate("A {PI} and a {PIE} are different", resourcePool);
+    assertTrue("No partial substitution", "A {PI} and a {PIE} are different".equals(result));
+  }
+
+  @Test
+  public void substitutionAndEscapeMixed() {
+    String result = interpreter.interpolate("A {PI} is not a {{PIE}}", resourcePool);
+    assertTrue("Substitution and escape mixed", "A 3.1415 is not a {PIE}".equals(result));
+  }
+
+  @Test
+  public void unbalancedBracesOne() {
+    String result = interpreter.interpolate("A {PI} and a {{PIE} remain unchanged", resourcePool);
+    assertTrue("Unbalanced braces - one", "A {PI} and a {{PIE} remain unchanged".equals(result));
+  }
+
+  @Test
+  public void unbalancedBracesTwo() {
+    String result = interpreter.interpolate("A {PI} and a {PIE}} remain unchanged", resourcePool);
+    assertTrue("Unbalanced braces - one", "A {PI} and a {PIE}} remain unchanged".equals(result));
+  }
+
+  @Test
+  public void tooManyBraces() {
+    String result = interpreter.interpolate("This {{{PI}}} remain unchanged", resourcePool);
+    assertTrue("Too many braces", "This {{{PI}}} remain unchanged".equals(result));
+  }
+
+  @Test
+  public void randomBracesOne() {
+    String result = interpreter.interpolate("A {{ starts an escaped sequence", resourcePool);
+    assertTrue("Random braces - one", "A {{ starts an escaped sequence".equals(result));
+  }
+
+  @Test
+  public void randomBracesTwo() {
+    String result = interpreter.interpolate("A }} ends an escaped sequence", resourcePool);
+    assertTrue("Random braces - two", "A }} ends an escaped sequence".equals(result));
+  }
+
+  @Test
+  public void randomBracesThree() {
+    String result = interpreter.interpolate("Paired { begin an escaped sequence", resourcePool);
+    assertTrue("Random braces - three", "Paired { begin an escaped sequence".equals(result));
+  }
+
+  @Test
+  public void randomBracesFour() {
+    String result = interpreter.interpolate("Paired } end an escaped sequence", resourcePool);
```

Comment:
could you add some tests for multiline string - and paired/unpaired  `{` and `{{` or `}` `}}`

Reply:
eg.
```
this is a {
test}
```

or

```
this is a
test }}
```

Reply:
Have expanded the set of tests for multi-line - both success and failures. There are now 18 unit tests.